README
		     @(#)README	1.22	2/29/96

			   sybperl 2.0
	
	  This is sybperl, the Sybase extensions for Perl5.

       Copyright (c) 1991,1992,1993,1994,1995  Michael Peppler
			 All rights reserved.
			 
  This program is free software; you can redistribute it and/or modify
  it under the same terms as Perl itself.

  You should have received a copy of the Perl license along with Perl;
  see the file README in Perl distribution.
 
  You should have received a copy of the GNU General Public License
  along with Perl; see the file Copying.  If not, write to the Free
  Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

  You should have received a copy of the Artistic License along with
  Perl; see the file Artistic.

---------------------------------------------------------------------

** NOTE: This version requires Perl 5.002 **


Please read the UPGRADE file if your previous version of sybperl was
1.0xx.

Please read the CHANGES file if you are upgrading from a previous
version of sybperl 2.x.

There are four modules in this distribution: Sybase::DBlib,
Sybase::BCP, Sybase::Sybperl and Sybase::CTlib.

Sybase::DBlib implements the Perl 5 version of the DB-Library
API. This differs slightly from the Perl 4 version. Note that this is
not a full implementation of the DB-Library API.
Sybase::BCP is a simplified bulk copy API, and requires Sybase::DBlib.
The Sybase::Sybperl module implements the sybperl 1.0xx API, and is
built on top of Sybase::DBlib.

Sybase::CTlib implements a subset of the newer Client Library API,
and should still be considered 'beta' software, although it is rapidly
becoming my favorite API for Perl/Sybase programing. Client Library is
included with Sybase's OpenClient libraries as of System 10.

---------------------------------------------------------------------

Installation:

Before you start, you have to consider the following issues:

a) Is Perl 5.002 (or later) already installed at your site?

   If Perl 5.002 (or later) is the default production version of Perl
   at your site, then I recommend that you unpack the sybperl
   distribution outside of the Perl source tree. The sybperl build
   process will normally be able to find the pieces it needs if this
   is the case, and it will be easier for you to manage upgrades to
   Perl and/or to the add-on extensions that you may have if they are
   kept separate.

   If 5.002 (or later) is not yet the default version of Perl, or has
   not been made available via an alternate path, then you should
   unpack the sybperl distribution in the ext/ directory of the Perl
   5.002 source tree. NOTE: this method is not well tested, and you
   may run into problems (unfortunately).

b) Do you want the Sybase modules to be dynamically loadable?

   Most Unix systems come with the capability of loading object code
   into an already executing program dynamically, usually via some
   sort of special library (on SunOS, for example, it's libdl.so.1.0).

   The advantage of dynamic loading is that you do not need to install
   a new Perl binary to provide new functionality, and the size of the
   binary does not grow as you add this functionality. The cost is a
   small performance penalty on startup.

   Static loading might be more advantageous to those specific modules
   which you know you will be using very often (however, I use dynamic
   loading here at ITF, although nearly all our Perl scripts use the
   Sybase modules).

   Note that if you are using HP-UX, you can only use dynamic loading
   if you have the add-on (extra cost) C compiler (or GNU gcc) and if
   you have the System 10 shared libraries (libsybdb.sl vs libsybdb.a)
   as HP-UX can't load an object dynamically which has not been
   compiled with the PIC (position independant code) flag.

c) The DO_TIE compile option

   The Sybase::DBlib and Sybase::CTlib modules define a number of
   attributes that are accesible via the database handle:

       $dbh = new Sybase::CTlib $user, $passwd...
       $dbh->{UseDateTime} = CS_TRUE;

   When the DO_TIE compile option is set, these attributes get 'tied'
   into a special package, thus enabling syntax verification, so that
   errors like

       $dbh->{UseDateTIme}

   will be caught.

   The downside is that you won't be able to add your own attributes
   to the $dbh on the fly. You must pre-declare any additional
   attributes you wish to use by passing them as a hash reference in
   the new()/dblogin()/ct_connect() calls (see the documentation and
   the Sybase::BCP module for examples). The DO_TIE option is set via
   the EXTRA_DEFINES option of the CONFIG file, and is enabled by
   default.
   
d) The -L/usr/local/lib switch and libtcl.{a, so, sl}

   Perl's Configure script adds this switch to the normal link or load
   line used when building extensions. This is fine, but if you have
   installed tcl/tk at your site, there is a strong chance that tcl's
   libtcl.a library has been placed in /usr/local/lib. Sybase's Client
   Library also includes a library called libtcl.{a,so,sl}. If the
   -L/usr/local/lib switch is present on the link command line then
   the wrong version of libtcl will be used, and the Sybase::CTlib
   module will not work. The solution is to either move or rename
   tcl/tk's libtcl.a while you build sybperl, or to remove the
   -L/usr/local/lib flag from Perl's Config.pm file.
   
Once you have answered these four questions, you can proceed:

   1) Unpack the package.
      See comments under a) above.
      
      Note that even if you don't want to install Perl 5.002 as a
      production version right now, you can still install it in some
      other place, and access that version by setting your PATH
      variable accordingly (I have a production version Perl5.002
      under /usr/local/{bin,lib} and a debug version of Perl5.002
      under /f2/{bin, lib} and change my path when I need to test
      things with debugging turned on).
      
   2) Edit CONFIG
      See comments under b) above regarding dynamic vs static loading.
      Some systems require extra libraries to build (Solaris 2.x
      requires -lnsl, for example). You should check your Sybase
      OpenClient Supplement for your OS and add any special libraries
      to the EXTRA_LIBS line. The required information can be found on
      the web at http://cobweb.sybase.com:8000/con10022/tctsg (you'll
      need to dig around a bit, though), or via AnswerBase.

      The DBLIBVS and CTLIBVS variables need to be set correctly for
      the version of OpenClient that is available at your site. Below
      is a table which lists which versions of Open Client were
      shipped with the different Sybase Server releases:

      Sybase version		   DBLIBVS   CTLIBVS
      - 4.0			       400	   0
      - 4.2 (includes MSSQL 4.2)       420         0
      - 4.8, 4.9.x                     461         0
      - 10.0.x                        1000       100

      You can check for the DB-Library version by looking at the top
      of the sybdb.h include file (it should be located in
      $SYBASE/include/sybdb.h, where $SYBASE is your Sybase
      installation directory.)

      Determine if you need any additional #define's for the
      EXTRA_DEFS line: With Perl 5.001m you needed -DUNDEF_BUG, but
      this bug has been corrected in 5.002, and you can also add
      -DDO_TIE to get the stricter attribute checking behavior (see
      point c. above).

      Uncomment the LINKTYPE line if required (see point b. above).

   3) Type: perl Makefile.PL
      
   4) Type: make

      If you get errors (such as undefined variables) it may be
      because your DBLIBVS or CTLIBVS does not correspond to the Open
      Client versions that are in fact installed at your site. Check
      them again. In the case of the Sybase::DBlib module, it doesn't
      hurt to build with DBLIBVS set to 400 as this is pretty well
      guarateed to work (provided you have *some* version of
      DB-Library!)

   6) Edit PWD
      Enter the user id and password that you want to use to run the
      tests. It's probably a good idea to reset this file to its
      original state once you're done, unless you really don't have
      any security concerns!

      The tests are non-destructive (ie only SELECT statements),
      except for Sybase::BCP which creates a temp table (#bcp) to test
      the bulk copy features.
      
   6) Type: make test
      If you decided to use static loading then a new perl binary will
      be generated at this point.

      If you get undefined symbols when testing CTlib (such as _net_*
      or _intl_getenv not defined), then I'd bet that you have either
      a libtcl.* or libintl.* in /usr/local/lib.  See point d) above.
      
   7) If everything went OK, type: make install
      Note that the man page does not get installed automatically. To
      install the documentation, cd to the pod directory, and read the
      README file there.
      Also note that even if you've asked for static linking and a
      perl binary has been built in the sybperl source directory, no
      perl binary gets installed (see step 8).

   8) If the LINKTYPE is static type:

	 make inst_perl MAP_TARGET=perl

      to install the new Perl binary. This step is not required when
      using dynamic linking as the existing Perl binary will load the
      new code dynamically when required.

   9) Read the manual :-)

      Look for it in the pod/ directory. You can create .man or .html
      versions from the .pod file by using pod2man or pod2html, both
      of which come with Perl.
      The manual may not be perfect yet, but it comes to 16 PostScript
      pages here, which is "respectable", I guess.


In case of problems:
====================

First, read the FAQ file in the sybperl distribution.

Second, fire off an e-mail to me (mpeppler@itf.ch) or to the Sybperl
mailing list (sybperl-l@trln.lib.unc.edu) with a *full* description of
what doesn't work, including version numbers of Perl and sybperl, the
hardware and OS version, and the Sybase Open Client version you are
using.

The regression tests don't attempt to test *everything*, but they do
test all of the basic functions, and should work cleanly.

There are quite a few small example scripts in eg/ which illustrate
the use of both the DBlib and CTlib modules.

There is a mailing list for discussing sybperl related questions
(sybperl-l@trln.lib.unc.edu). Send a message to
sybperl-l-request@trln.lib.unc.edu with 'subscribe your name'
in the body to subscribe to the list.



Reporting Bugs:
===============

I have set up a bug tracking database at ITF, to which bugs can be
submitted via e-mail. If you have a comment, bug report or feature
request you would like to make please use the template in the SPR file
in this directory.

When submitting bugs, or any sort of problem report, please make sure
you include full version information for Sybperl, Perl, Sybase and
also what hardware/OS configuration you are on. The exact Sybperl
versions can be extracted by running the 'Version' script in this
directory. Perl's configuration can be summarized with the 'myconfig'
script in the Perl source directory.


Comments, criticism, et.al. VERY welcome!

Michael Peppler - mpeppler@itf.ch	Phone: +4122 818 3021
ITF Management SA			Fax:   +4122 312 1322



CHANGES
	$Id: CHANGES,v 1.22 2005/03/20 19:50:59 mpeppler Exp $

2.18    Bug Fixes

        Various BLK fixes.
        Fixed Sybase::BCP and Sybase::BLK tests which failed against
        12.5.3 servers (changes on the Server side).

	Bugs Fixed:

	577 - perl Makefile.PL fails if umask is 0.

2.17    Kerberos Network login

	Support for CS_SEC_NETWORKAUTH and CS_SEC_SERVERPRINCIPAL
	connection properties in Sybase::CTlib.
	Support CS_SERVERADDR connection properties in Sybase::CTlib,
	and also allow the $server parameter to ct_connect to be
	specified as "host:port" (requires OpenClient 12.5.1 or later).
	Some documentation changes/improvements in Sybase::CTlib.

	Bug Fixes
	546	umask checking doesn't work on Windows.
	555	blk_rowxfer doesn't handle datatype conversion errors.
	560	Long binary columns are truncated by ct_fetch.
	542	Fetch binary in raw format.
	538     DBlib doesn't export TRACE_xxx symbols.

2.16    Various bug fixes

        Added DBSETLENCRYPT() to Sybase::DBlib.
	Added DBDEAD() to Sybase::Sybperl.
	Rearranged the directory structure somewhat so that all the
	tests are now in one directory and are run as one unit.
	64-bit builds should now be detected automatically (assuming
	the perl binary was built in 64-bit mode)
        Split out the documentation so that there are now pod entries
	in DBlib.pm and CTlib.pm.

	Bug Fixes
	494   Do not try to use the ocs.cfg file if it exists.
        499   Change MAXNAME to DBMAXNAME in the DBlib source.
        500   $date->info return invalid data.
        511   Change login to syb_login in DBlib.xs (work-around for FreeBSD).
	516   Money/Numeric values lose precision when not using 
              UseMoney/UseNumeric.
	518   looking for non-existent library file (solaris 2.8/ASE 12.5.0.3
	      perl 5.8)
	498   Problem compiling sybperl under HP-UX 64-bit mode.
	488   Sybase::CTlib and DBD::Sybase end up using the same error 
	      handler
	470   BLK.pm does not handle empty columns as NULL, and does not 
              have retry logic of BCP.pm
	410   BCP.pm has regex bug.
	379   Sybase::BCP::config() does not correctly handle incremental 
              changes

2.15    Various bug fixes

        Support for 64bit OpenClient on platforms that support this
        (Thanks to Pamela Schulz for her help with this)
        Added CTlib/t/xblk.t and CTlib/t/prepare.t test scripts.
	CS_BLK_CANCEL was not exported.
	For building with FreeTDS you can now add -DNOBLK to the 
	EXTRA_DEFS line in CONFIG.

	Bug Fixes
	468   Makefile.PL has syntax error when used with perl 5.8.x
	460   Idem.
	439   Makefile.PL fails for very old Sybase installation (i.e.
              installation with no CT-Library)
	423   Fix Makefile.PL for Win32 builds.
	422   t/bcp.t fails against 11.0.3.3 linux servers.


2.14	And more (minor) bug fixes

	config.pl's getExtraLibs() now tries to determine the Client
	Library version to avoid linking with -linsck or -ltli for
	OC 11 or later (as the correct library is loaded dynamically 
	based on the values defined in $SYBASE/config/libtcl.cfg)
        Added support for CS_HAFAILOVER connect option (thanks to 
	Ranko Popovic)
        Added CTlib version of nsql().
	Documentation patches from Geoffrey Rommel.
	Fixed memory leak when using UseDateTime/UseMoney/UseNumeric.
	Add Async calls for CTlib (ct_poll(), etc).
	Add CS_DATA_LCHAR connection capability.
        Add ct_dyn_prepare() and ct_dyn_execute() to handle "dynamic SQL"
	- thanks to Masatsuyo Takahas for the patch.
	Add experimental blk_*() calls.
	Ask for the PWD entries during the perl Makefile.PL phase.
	Add Sybase::BLK module (based on Sybase::BCP) (*many* thanks to
        Scott Zetlan for this!)
	
	Bug Fixes
	309   Sybase::BCP fails test if server is set to use 
	      datarows locking by default.

2.13    Yet more bug fixes

        Correct EMBED_SYBASE handling (bug in config.pl script).
	Fix building for Sybase 12.0 (uses SYBASE_OCS environment 
	variable.)
	Added dbretlen(), dbretname(), dbrettype() to DBlib.
        Minor changes to support wide data in Sybase 12.5.

	340: Sybase::BCP fails unless REORDER is defined.
	342: DateTime calc method microseconds aren't.

2.12    More bug fixes

	Better EMBED_SYBASE, thanks to Mike Hiller.
	Added UseBix0x mode to CTlib.

	269: dbsqlsend() missing from Sybperl.pm emulation layer.
	315: -ldnet_stub directive in CONFIG ignored for OSF1/Tru64 Unix.
	312: DBINTERRUPT misspelled in DBlib.xs.

2.11	Various bug fixes.

	nsql() patch from W. Philipp Moore.
	Added EMBED_SYBASE configuration variable.
	dbreadtext() could cause a core dump if the internal call failed.
	Figured out what was needed to get Sybase::DBlib to talk to an 
	MS-SQL 7 dataserver. Sybase::CTlib still does not work with the
	normal, non-patched MS-SQL 7 server.
	Fixed bug 261: Money/Numeric potentially lose precision in
		       Sybase::CTlib.

2.10_02
	Fix incorrect dropping of locale structure in CTlib::DESTROY.
	Link pod/sybperl.pod to CTlib/CTlib.pod and DBlib/DBlib.pod so 
	that perldoc Sybase::DBlib now works.

2.10_01

	Added dbbylist().
	Added DBIORDESC() and DBIOWDESC() (contributed by David Worenklein)
	Modifications to nsql() contributed by W. Phillip Moore.
	Added minimal support for languages and charsets to Sybase::CTlib.

2.10    CTlib TEXT/IMAGE handling

	**WARNING**

	Please see the sybperl.pod file for dbpoll() syntax
	changes.

	Added ct_get_data(), ct_send_data(), ct_data_info().
	Added some missing documentation.
	Added Sybase::CTlib::DBDEAD().
	Added Sybase::DBlib::dbcomputeinfo().

	Changed DBlib::dbpoll() to allow polling on a single handle.

	W. Philipp Moore supplied a patch to CTlib/Makefile.PL to 
	link with the Sybase _r (re-entrant) libraries if perl
	was compiled with threading enabled. **THERE IS NO GUARANTEE
	THAT A THREADED VERSION OF Sybase::CTlib WILL WORK CORRECTLY!!!***
	Please read README.thread for details.

	Bug Fixes:

	247: Sybase::DBlib::sql() returns invalid results in $doAssoc mode

Makefile.PL
# -*-Perl-*-
#	$Id: Makefile.PL,v 1.14 2004/11/19 10:35:37 mpeppler Exp $
require 5.002;
use ExtUtils::MakeMaker;
use Config;

use strict;

require './util/config.pl';

my $sattr = &config();
my $written_pwd_file = 'PWD';

configPwd();

my $linktype = defined($$sattr{LINKTYPE}) ? $$sattr{LINKTYPE} : 'dynamic';

#if($Config{lns}) {
#    system("cd CTlib; $Config{lns} ../pod/sybperl.pod CTlib.pod") if(!-e 'CTlib/CTlib.pod');
#    system("cd DBlib; $Config{lns} ../pod/sybperl.pod DBlib.pod") if(!-e 'DBlib/DBlib.pod');
#}
    


foreach (@INC) {
    next if(/site_perl/);
    next if(/blib/);
    if(-d "$_/Sybase") {
	warn <<"MSG_EOF";
*** WARNING ***
An installed copy of the Sybase modules found in $_
This can cause a version conflict as this version installs in
    $Config{installprivlib}/site_perl
which comes later in Perl $]'s include directory list.

You may need to manually remove these older Sybase directories if
sybperl $sattr->{VERSION} does not run correctly.

One way to do so would be to run the following command
    mv $_/Sybase $_/Sybase.old

MSG_EOF
#'	
    }
}

WriteMakefile('DISTNAME' => "sybperl",
	      'NAME' => 'Sybase',
	      'VERSION' => $$sattr{VERSION},
	      'dist' => {'TARFLAGS' => "cvf", 'COMPRESS' => "gzip"},
	      'clean' => { FILES => $written_pwd_file },
	      'LINKTYPE' => $linktype,
	      'XSPROTOARG' => '-prototypes',
	      ($] >= 5.005 ?
	       (ABSTRACT   => 'Sybase API modules',
		AUTHOR     => 'Michael Peppler (mpeppler@peppler.org)') : ()),
	      ($] >= 5.005 && $^O eq 'MSWin32' && 
	       $Config{archname} =~ /-object\b/i ? (CAPI => 'TRUE') :()),
              'MAN3PODS' => { 'pod/sybperl.pod' => 'blib/man3/sybperl.3' }
	      );



sub MY::libscan
{
    my($self, $path) = @_;
    return '' if $path =~ m:/SCCS/:;
    $path;
}

sub configPwd {
    open(IN, "PWD.factory") || die "Can't open PWD.factory: $!";
    my %pwd;
    while(<IN>) {
	chomp;
	next if(/^\s*\#/);
	next if(/^\s*$/);
	my ($key, $val) = split(/=/, $_);
	$pwd{$key} = $val || "undef";
    }
    close(IN);
    select(STDOUT); $| = 1;

    print "The sybperl modules need access to a Sybase server to run the tests.\n";
    print "To clear an entry please enter 'undef'\n";
    print "Sybase server to use (default: $pwd{SRV}): ";
    $pwd{SRV} = getAns(0) || $pwd{SRV};
    print "User ID to log in to Sybase (default: $pwd{UID}): ";
    $pwd{UID} = getAns(0) || $pwd{UID};
    print "Password (default: $pwd{PWD}): ";
    $pwd{PWD} = getAns(1) || $pwd{PWD};
    print "Sybase database to use on $pwd{SRV} (default: $pwd{DB}): ";
    $pwd{DB} = getAns(0) || $pwd{DB};

    warn "\n* Writing login information, including password, to file $written_pwd_file.\n\n";
    # Create the file non-readable by anyone else.
    my $old_umask;
    unless($^O =~ /MSWin32/) {
	$old_umask = umask(077);
	warn "cannot umask(): $!" unless defined($old_umask);
    }
    open(OUT, ">$written_pwd_file") || die "Can't open $written_pwd_file: $!";
    unless($^O =~ /MSWin32/) {
	umask($old_umask) != 077 && warn "strange return from umask()";
    }
    print OUT <<EOF;
# This file contains optional login id, passwd and server info for the test
# programs:
# You probably don't want to have it lying around after you've made
# sure that everything works OK.

EOF
    foreach (keys %pwd) {
	$pwd{$_} = '' if $pwd{$_} eq 'undef';
	print OUT "$_=$pwd{$_}\n";
    }
    close(OUT);
}

sub getAns {
    my $flag = shift;

    if($flag && -t) {
	eval {
	    require Term::ReadKey;
	    Term::ReadKey::ReadMode('noecho');
	};
    }
    my $ans = <STDIN>;
    if($flag && -t) {
	eval {
	    Term::ReadKey::ReadMode('normal');	# reset terminal
	};
    }
    $ans =~ s/^\s+//;
    $ans =~ s/\s+$//;

    return $ans;
}

    
